File: 2024/2.add-two-numbers.py
#
# @lc app=leetcode id=2 lang=python3
#
# [2] Add Two Numbers
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        carry = 0
        node1 = l1
        node2 = l2
        dummy = ListNode()
        result_node = dummy
        while node1 is not None or node2 is not None:
            result_node.next = ListNode()
            result_node = result_node.next
            val = (node1.val if node1 else 0) + (node2.val if node2 else 0) + carry
            result_node.val = val % 10
            carry = val // 10
            node1 = node1.next if node1 else node1
            node2 = node2.next if node2 else node2
        if carry > 0:
            result_node.next = ListNode()
            result_node = result_node.next
            result_node.val = carry
        return dummy.next

# @lc code=end

File: 2024/2.add-two-numbers.ts
/*
 * @lc app=leetcode id=2 lang=typescript
 *
 * [2] Add Two Numbers
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function addTwoNumbers(l1: ListNode | null, l2: ListNode | null): ListNode | null {
  let node1 = l1;
  let node2 = l2;
  const dummy = new ListNode();
  let root = dummy;
  let carry = 0;
  
  while (node1 !== null || node2 !== null || carry > 0) {
    const val = ((node1?.val ?? 0) + (node2?.val ?? 0) + carry) % 10
    carry = Math.floor(((node1?.val ?? 0) + (node2?.val ?? 0) + carry) / 10)
    const node = new ListNode(val)
    root.next = node;

    root = root.next;
    node1 = node1 ? node1.next : node1;
    node2 = node2 ? node2.next : node2;
  }

  return dummy.next;
};
// @lc code=end

